### DISCUSSION TITLE
m.tarenskeen’s PSS-480/-580 panel - update on mouse release?

### DISCUSSION THREAD

**Discussion by Kidwho** [RomanKubiak/ctrlr#411] [Q&A] [answered]
Hi all,

First of all, please note I have almost no experience creating or editing Ctrlr panels beyond a little dabble here and there, so please be kind :) I’m a bit user of Martin Tarenskeen’s fantastic panel for the PSS-580:

http://ctrlr.martintarenskeen.nl/

One functionality is that the synth accepts updates of the entire patch rather than individual parameter changes, so Martin has included a ‘Put’ button which you need to press after making an edit for the changes to transmit. I’m wondering if there is a way in Ctrlr to make this happen immediately and automatically on mouse release? i.e. you click a dial, make the change and when you release the mouse the ‘Put’ action is executed automatically?

Of course, true real time editing is not possible, and the unit’s MIDI buffer is easily clogged, so timing glitches etc are expected but this would at least reduce one step and make editing much faster.

I am in touch with Martin directly and will of course relay any suggestions or changes to him!

Thanks in advance for your help :)

**Comment by dnaldoog** [comment, marked as answer]
Hi @Kidwho!

Welcome to Ctrlr.

I added that function that fires when you press the _PUT_ button to be fired by some of the dials on the panel! I have no idea whether that is a good thing for MIDI messaging (is it too much for the machine? etc), but will give you an idea how to do that! I also added the _pssPut_ function to the EGPaint function that paints the graph.

You can see where to add a callback function on the right side of the picture. When _Fine Detune_ uiSlider is turned it will fire the same function that PUT does!

[PSS-580 (all fire pssPut)a_1_0_20210322_2022-01-19_23-16.zip](https://github.com/RomanKubiak/ctrlr/files/7897817/PSS-580.all.fire.pssPut.a_1_0_20210322_2022-01-19_23-16.zip)

![pssPut](https://user-images.githubusercontent.com/76831683/150149497-0b59f846-8e58-4a95-bca6-f0bd75634569.png)

**Reply by Kidwho** [reply]
Oh wow! I’ve just tested it out adding the ‘pssPUT’ function to a number of the knobs and it works amazingly well - thank you SO much @dnaldoog !!

There is indeed some glitching while the parameters are changing but that was always going to be the case on this machine, this just makes editing so much more enjoyable, fast and most importantly, creative.

Can’t wait to finish work today, set up the function on all the controls and have a proper play with it 🙂 I cannot thank you enough!! 🙏🙏

**Reply by Godlike-Productions** [reply]
To reduce glitching I ended up writing my own MIDI scheduler that runs on a timer callback (I use 200ms).  

Whenever I move a controller I set a flag bufferChanged to true.  In the timer callback I sample the values of all of the modulators every pass (ie once every 200ms).  If bufferChanged is true then it sends the MIDI message and copies the buffer into lastBuffer (so I can compare current buffer to lastBuffer (if they are different, set bufferChanged to true).

I manually start this process to put the panel into control mode, mainly so that I can turn MIDI traffic off when I'm doing bank and patch management, but if the synth processor can handle it, you could keep it on all the time.

The Jupiter 8 is easy to crash if you send it MIDI data too fast, so I've had to implement this, but I use it even on more modern synths to minimize sending unnecessary MIDI data.  I tend to write my own MIDI scheduler to do things like take advantage of running messages to make the panel as responsive as possible.

**Reply by dnaldoog** [reply]
> Oh wow! I’ve just tested it out adding the ‘pssPUT’ function to a number of the knobs and it …

Excellent to hear! 😀

**Comment by dnaldoog** [comment]
Hi @Kidwho!

I'm back!

There is actually a slightly more elegant and efficient way of doing this:

First you create a function that fires when any modulator changes its value (See __`allModFire`__ function in  image). 
In this function you write a lua table that checks whether the modulator name that fired that function is in the list. If it is in the list the `pssPut` function is fired, otherwise nothing happens.

I you don't want a _modulator_ to fire off the `pssPut` function, in the `set{}` table, either set it to false or comment it out with `--` or delete it from the  table altogether.


See attached: 
<pre>
--
-- Called when any modulator changes on the panel
--
-- @program modulator that changed
-- @value new modulator value
--

allModFire = function(--[[ CtrlrModulator --]] modulator, --[[ number --]] value)
local mName=L(modulator:getName())

local set={
    BANK = true,
    BB = true,
    FB = true,
    PMS = true,
    AMS = true,
    VDT = true,
    VIB = true,
    SUS = true,

    -- FM CARRIER PARAMETERS
    MUL_C = true,
    DT1_C = false, -- Won't fire pssPut()
  --  DT2_C = true, -- Won't fire pssPut()
    TL_C = true,
    LKSHI_C = true,
    LKSLO_C = true,
    RKS_C = true,
    AR_C = true,
    D1R_C = true,
    D2R_C = true,
    D1L_C = true,
    RR_C = true,
    SRR_C = true,
    AME_C = true,
    SIN_C = true,

    -- FM MODULATOR PARAMETERS
    DT1_M = true,
    DT2_M = true,
    MUL_M = true,
    TL_M = true,
    LKSHI_M = true,
    LKSLO_M = true,
    RKS_M = true,
    AR_M = true,
    D1R_M = true,
    D2R_M = true,
    D1L_M = true,
    RR_M = true,
    SRR_M = true,
    AME_M = true,
    SIN_M = true,
}
if set[mName] then
pssPut()
end
end
</pre>

Regards.

![pss580(2)](https://user-images.githubusercontent.com/76831683/150296498-0058ab82-f9fe-461b-8557-c0d8016074b2.png)

[PSS-580 (global midi send 2)_1_0_20210322_2022-01-20_16-43.zip](https://github.com/RomanKubiak/ctrlr/files/7903256/PSS-580.global.midi.send.2._1_0_20210322_2022-01-20_16-43.zip)

**Comment by mtarenskeen** [comment]
I am the original author of this panel, and following this thread.
I like the idea of a timer callback to reduce the datastream. But still I would prefer if it were possible to trigger pssPut only on mouse release, not when changing a value while holding the mouse button. For example if I move a button from value 0 to 63 and back to to value 1, I would like that the panel only triggers pssPut once after releasing the mouse, sending the data with this parameter changed from value 0 to value 1 only, instead of sequentially sendiing 127 x 72 bytes. 

If this can't be done I would like to try a timer callback function that checks every xxx (milli)seconds if any of the parameters in the patch have changed, and if so, fire pssPut.

**Comment by dnaldoog** [comment]
I just found a way to send MIDI on mouse release.

I create a component in a separate layer below a _uiSlider_ that has a _onMouseMove_ function which when triggered sends the MIDI after the user moves the mouse away from the _uiSlider_. 

But it only sends once because each _uiSlider_ sitting above the _uiComponent_ sets a flag to `true` which is then immediately reset to `false` by onMouseMove after one MIDI send when the mouse exits the _uiSlider_ space (the _uiSlider_ sitting on top of the component blocks any of the uiComponent's_ mouse callbacks). 

Possibly no need for timers or much lua code at all! Seems to work very well.

<pre>
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
setFlag = function( mod, value, source)
sendFlag=true
end
</pre>

<pre>
--
-- Called when the mouse moves over a component
--

mouseMove = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
      if sendFlag == true then
        local m1 = panel:getModulatorByName("m1"):getModulatorValue()
        local m2 = panel:getModulatorByName("m2"):getModulatorValue()
        local sysex = string.format("F0 41 00 23 %.2x %.2x F7", m1, m2)
        panel:sendMidiMessageNow(CtrlrMidiMessage(sysex))
        sendFlag = false
    end
end

</pre>

![mouseover](https://user-images.githubusercontent.com/76831683/150334674-4f012285-0b83-4fec-81ad-5b272a325c7e.png)

[Send Bulk Midi on Mouse Move_1_0_2022-01-20_20-50.zip](https://github.com/RomanKubiak/ctrlr/files/7904997/Send.Bulk.Midi.on.Mouse.Move_1_0_2022-01-20_20-50.zip)

**Reply by dobo365** [reply]
Very interesting, John! Will test that!
I was not really in search of an OnMouseRelease event trigger but after some thoughts, I guess I may need that sooner than expected ;-)

**Reply by dnaldoog** [reply]
Thanks Dominique! I tried it briefly on the panel in question but it didn’t work too responsively. Maybe something to do with uiImageSlider as opposed to vanilla uiSlider or that with modulators packed closely together there would need to be enough space for the mouse move to be detected. Still, could be very useful in some situations!

**Reply by Godlike-Productions** [reply]
I vaguely recall that I had some issues with mouseUp on the Kurzweil panel when programming the data wheel. I ended up with some kind of hack to get it to track how far we'd turned it and then send the value when we released it.  We programmed it as a sprung controller though (because ctrlr doesn't really do endless encoders). I can't remember exactly what the issue was, but I can go an check the code if anyone wants to know how we did it.

**Comment by mtarenskeen** [comment]
Hi guys,

Thanks for all suggestions.
I have released a new version of my PSS panels for testing.

I now use a timer callback, and re-use much of the LUA code that already was available in my panel. Didn't need much extra work and lines of code this way. 
 I used existing pssCopy() to remember the latest state of the modulators.
When I change a modulator every 1000ms I check the current state of the modulators with the last pssCopy state. If anything has changed pssPut() is called and pssCopy state is updated.

Have not done heavy testing yet, but seems to work.

Latest version 20220127 can be download for testing from my website [http://ctrlr.martintarenskeen.nl](url)

**Reply by Kidwho** [reply]
Thanks so much everyone for all the hard work on this! What an awesome and friendly community 🙂 @mtarenskeen trying to download the beta to test out but can’t see it on the site? Thanks!

**Comment by mtarenskeen** [comment]
My wrong. Please try again, it should be visible now on 
http://ctrlr.martintarenskeen.nl <http://ctrlr.martintarenskeen.nl/>

Op donderdag 27 januari 2022 om 04:48:28 -0800 schreef Kidwho 
***@***.***>:

**Comment by dnaldoog** [comment]
@mtarenskeen That's a nice solution, but could you not just do this?

<pre>
function pssAutoPut()
	-- This variable stops index issues during panel bootup
	if panel:getRestoreState() == true or panel:getProgramState() == true then
		return
	end
	timer:setCallback(1, timerCallback)
	timer:startTimer(1,1000)
end

function timerCallback()
		pssPut()
		--pssCopy() does this function do anything?
		timer:stopTimer(1)
end
</pre>

... instead of 

<pre>
function pssAutoPut()
	-- This variable stops index issues during panel bootup
	if panel:getRestoreState() == true or panel:getProgramState() == true then
		return
	end

	timer:setCallback(1, timerCallback)
	timer:startTimer(1,1000)
end

function timerCallback1()
	local b = bank:getValue()
	local current = panel2data(b)

	CHANGED=false
	for i = 1,33 do
		if current[i] ~=editbuffer[i] then
			CHANGED=true
			break
		end
	end

	if CHANGED == true then

for i=1,33 do
console(""..current[i])
end
		pssPut()
		pssCopy()
	--	console("AUTOPUT SUCCEEDED?")
		timer:stopTimer(1)
	end
end
</pre>

**Reply by mtarenskeen** [reply]
Delayed reaction ...
You are right. Sometimes the simpler solution is the better solution. It works even better now, because it doesn't break the copy/paste functionality anymore.

Now this is used when a modulator value changes in latest version 20220225 that can be downloaded from [http://ctrlr.martintarenskeen.nl](url) :
```

function pssAutoPut()
	-- This variable stops index issues during panel bootup
	if panel:getRestoreState() == true or panel:getProgramState() == true then
		return
	end
	timer:setCallback(1, timerCallback)
	timer:startTimer(1,1000)
end

function timerCallback()
		pssPut()
		-- console("AUTOPUT SUCCEEDED?")
		timer:stopTimer(1)
end

```
